### PR TITLE
Kernel/USB: Add a rudimentary interrogation only EHCI driver

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
~~### Kernel: Add a `void` specialization to TypedMapping~~

### Kernel/USB: Add a rudimentary interrogation only EHCI driver

This adds a simple EHCI driver that currently only interrogates the
device and checks if all ports are addressable via associated legacy
controllers (companion controllers), and warns if this is not the case.

This also adds a lot of the other data structures needed for actually
driving the controller, but these are currently not hooked up to
anything.

To test this run with `SERENITY_EXTRA_QEMU_ARGS="--device usb-ehci"`
or the q35 machine type

### Meta: Hook up the UHCI HCs up to the EHCI HCs in the q35 machine

Now devices connected to the EHCI controllers show up on the hooked up
UHCI companion controllers.

---

CC: @supercomputer7

### CODE REVIEW

File: Kernel/Bus/USB/EHCI/EHCIController.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2023, Leon Albrecht <leon.a@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "./EHCIController.h"
```
Comment (supercomputer7) [inline Kernel/Bus/USB/EHCI/EHCIController.cpp:7]: why don't you use absolute path here?

File: Kernel/Bus/USB/EHCI/EHCIController.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2023, Leon Albrecht <leon.a@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include "./Registers.h"
```
Comment (supercomputer7) [inline Kernel/Bus/USB/EHCI/EHCIController.h:9]: We should use absolute paths.

**Comment by supercomputer7** [comment]
I don't really like the idea of using bit-fields in the code. It means we assume that we will never port the OS to big-endian CPU architecture, so if this is safe to do so them it might be OK...
Also please look at this - https://stackoverflow.com/questions/6043483/why-bit-endianness-is-an-issue-in-bitfields

**Comment by Hendiadyoin1** [comment]
> so if this is safe to do so them it might be OK...

As far as I can tell gcc and clang agree on the layout on little-endian systems*,** of bit-fields as little bit endian, so the 0th bit comes first and so on, which mostly means you have to read tables backwards in the ehci case
So I think it's safe enough for now

*: clang does not like bitcasts of bitfields in `static_asserts`, so I go by the precedent we have from WASM and other parts of the kernel
**: GCC on PPC64LE seems to assume big-endianness in constant expressions, not sure if that is intentional though

**Comment by Hendiadyoin1** [comment]
Screw you @kleinesfilmroellchen and your months old PR finally getting merged!

**Comment by Hendiadyoin1** [comment]
Will port it to run.py shortly

File: Kernel/Memory/TypedMapping.h [hunk 3]
```diff
@@ -28,6 +28,19 @@ struct TypedMapping {
     size_t length { 0 };
 };
 
+template<>
+struct TypedMapping<void> {
```
Comment (ADKaster) [inline Kernel/Memory/TypedMapping.h:32]: :T If you're using `void*`, then it's not exactly a *typed* mapping, is it? Is there really no overlapping type we can use here for EHCI? If there really isn't we should just use a volatile void* to show that the mapping is super unusual/annoying.

File: Kernel/Bus/USB/EHCI/EHCIController.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    return controller;
+}
+
+EHCIController::EHCIController(PCI::DeviceIdentifier const& pci_device_identifier, Memory::TypedMapping<void> mapped_registers)
+    : PCI::Device(pci_device_identifier)
+    , m_mapped_registers(move(mapped_registers))
+{
+    m_cap_regs = bit_cast<CapabilityRegisters const*>(m_mapped_registers.base_address().get());
+    m_op_regs = bit_cast<OperationalRegisters volatile*>(m_mapped_registers.base_address().get() + m_cap_regs->capability_length);
```
Comment (ADKaster) [inline Kernel/Bus/USB/EHCI/EHCIController.cpp:40]: Hm, like, can we map this set of regs beforehand? who is pulling random values from m_mapped_registers later, not through the cap/op regs?

File: Kernel/Bus/USB/EHCI/Registers.h [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+        u32 raw;
+    } port_status_control[];
+    static_assert(AssertSize<PortStatusControl, 32 / 8>());
+};
+// Table 2-8. Host Controller Operational Registers
+static_assert(__builtin_offsetof(OperationalRegisters, command) == 0x00);
+static_assert(__builtin_offsetof(OperationalRegisters, status) == 0x04);
+static_assert(__builtin_offsetof(OperationalRegisters, interrupt_enable) == 0x08);
+static_assert(__builtin_offsetof(OperationalRegisters, frame_index) == 0x0C);
+static_assert(__builtin_offsetof(OperationalRegisters, segment_selector) == 0x10);
```
Comment (ADKaster) [inline Kernel/Bus/USB/EHCI/Registers.h:278]: I wonder if we should start using that OFFSET_OF macro from AK/Types.h for new code. food for thought, not for changes though :)

**Review by ADKaster — COMMENTED** [review]
Looks reasonable, but I'm skeptical about `TypedMapping<void>`

**Review by ADKaster — APPROVED** [review]

**Pull request merged**